torch/export/_trace.py

```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
 
 
+def _skip_autograd_kernel(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    with torch._C._AutoDispatchBelowAutograd():
+        return kernel(*args, **kwargs)
+
+
+def _register_cia_to_meta(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(
+            torch._C.DispatchKey.CompositeImplicitAutograd, *args, **kwargs
+        )
+    else:
+        return NotImplemented
+
+
+@contextmanager
+def override_composite_implicit_decomp(ops_to_preserve):
+    # This function overrides CompositeImplicitAutograd decomp for
+    # functional composite ops. Ideally we want to not-decompose
+    # ALL composite ops but today's C++ functinalization relies on
+    # the fact that it is working with the opset after decomp is run.
+    # Hence we can only do it for functional ops. One caveat is that
+    # there are some composite ops that lie about their schema (claimed to be
+    # functional but not really aka dropout), for these cases, we just decompose.
+    saved_tables = {}
+    patched_ops = set()
+    for op_overload in ops_to_preserve:
+
+        def can_preserve(op_overload):
+            if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+                return False
+            if op_overload in FunctionalTensor.metadata_fns:
+                return False
+
+            alias_info = len(
+                [i for i in op_overload._schema.arguments if i.alias_info is not None]
+            )
+
+            is_mutating_or_aliasing = alias_info != 0 or op_overload._schema.is_mutable
+
+            if is_mutating_or_aliasing:
+                return False
+
+            if not torch._C._dispatch_has_kernel(op_overload.name()):
+                return False
+
+            if not torch._C._dispatch_has_kernel_for_dispatch_key(
+                op_overload.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+            ):
+                return False
+
+            return True
+
+        if not can_preserve(op_overload):
+            warnings.warn(
+                f"We can't preserve {op_overload} in export because this op is unsafe to keep"
+            )
+            continue
+
+        saved_tables[op_overload] = op_overload.py_kernels.copy()
+        patched_ops.add(op_overload)
+        for override_dispatch_key in [
+            torch._C.DispatchKey.AutogradCPU,
+            torch._C.DispatchKey.AutogradCUDA,
+            torch._C.DispatchKey.AutogradMeta,
+        ]:
+            if override_dispatch_key not in op_overload.py_kernels:
+                # conv1d, conv2d, conv3d don't work with default Autograd key
+                op_overload.py_impl(override_dispatch_key)(
+                    functools.partial(_skip_autograd_kernel, kernel=op_overload)
```

Comment:
this should be updated to use the autograd_error kernel that custom ops uses, right? (fyi @zou3519)

Reply:
Oh yeah i forgot to do it, will do.

Reply:
@tugsbayasgalan and I talked offline, and we don't think this is necessary: what situations are we trying to error out on? If we are exporting a module, then hoping that `exported_program.module().forward(x).backward()` will error out, then the operator needs to use the autograd_error kernel during the execution of the forward (and not during export)

Reply:
I agree that this won't help much if the user does the above. My thought was mostly that:

(1) "not decomposing a CompositeImplicit op" during tracing has safety issues if you happen to try to use the autograd engine during tracing (e.g. if you're tracing out the joint graph)

(2) if we're adding the capability to skip CompositeImplicit decomps for export now, we may as well add it in a safe way that will automatically error when someone eventually wants to get joint graph tracing to work (instead of having to remember the correctness conditions and figure out how to error properly later).

If this is easy to do (e.g. through the autograd_error API that custom ops uses), then it seems useful to do it right the first time. But lmk if you disagree / think this is a pain.

Reply:
I tried doing:
```
op_overload.py_impl(AutogradCPU)(make_autograd_impl(op_overload, Info(None, None)))
```
But ran into following error:
```
    flat_f_outs = f(*flat_f_args)
  File "/data/users/tmanlaibaatar/pytorch/torch/_functorch/_aot_autograd/utils.py", line 178, in flat_fn
    tree_out = fn(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 760, in functional_call
    out = PropagateUnbackedSymInts(mod).run(
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 146, in run
    self.env[node] = self.run_node(node)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/experimental/symbolic_shapes.py", line 5462, in run_node
    result = super().run_node(n)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 203, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 275, in call_function
    return target(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_ops.py", line 670, in __call__
    return self_._op(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_library/autograd.py", line 113, in autograd_impl
    result = forward_no_grad(*args, Metadata(keyset, keyword_only_args))
  File "/data/users/tmanlaibaatar/pytorch/torch/_library/autograd.py", line 40, in forward_no_grad
    result = op.redispatch(keyset & _C._after_autograd_keyset, *args, **kwargs)
TypeError: unsupported operand type(s) for &: 'FunctionalTensor' and 'torch._C.DispatchKeySet'
```

It seems like i am using this API kinda wrong. When i was reading the code, it looks like we are handling the last arg little differently https://github.com/pytorch/pytorch/blob/d52684e9a8b95f8b0d06f0bf08e6a39846cb3ae6/torch/_library/autograd.py#L44

@zou3519 any clues?

Reply:
There's two calling conventions for the Dispatcher. If you have an operator "sin(Tensor x) -> Tensor" -- you can register a kernel that accepts "(x: Tensor)" or a kernel that accepts a DispatchKeySet first: "(keyset: DispatchKeySet, x: Tensor)".

To use `make_autograd_impl`, we need to add support for the latter calling convention in `py_impl`. This shouldn't be hard:
- we'd add an extra argument to `py_impl` that is a bool flag for the calling convention and then store this information somewhere
- then we need to call into the py_impl with the calling convention.

You need to do this anyways to make the redispatch faithful (that is, the redispatch should use opoverload.redispatch).